TileStache/Goodies/VecTiles/transform.py

```diff
@@ -555,6 +556,111 @@ def _geom_dimensions(g):
+    'LINESTRING (0 0, 1 1)'
+    >>> _filter_geom_types(Polygon([(0,0),(1,1),(1,0),(0,0)],[]), _POLYGON_DIMENSION).wkt
+    'POLYGON ((0 0, 1 1, 1 0, 0 0))'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POINT(0 0), LINESTRING(0 0, 1 1))'), _POINT_DIMENSION).wkt
+    'POINT (0 0)'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POINT(0 0), LINESTRING(0 0, 1 1))'), _LINE_DIMENSION).wkt
+    'LINESTRING (0 0, 1 1)'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POINT(0 0), LINESTRING(0 0, 1 1))'), _POLYGON_DIMENSION).wkt
+    'GEOMETRYCOLLECTION EMPTY'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POINT(0 0), GEOMETRYCOLLECTION (POINT(1 1), LINESTRING(0 0, 1 1)))'), _POINT_DIMENSION).wkt
+    'MULTIPOINT (0 0, 1 1)'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (MULTIPOINT(-1 -1, 0 0), GEOMETRYCOLLECTION (POINT(1 1), POINT(2 2), GEOMETRYCOLLECTION (POINT(3 3))), LINESTRING(0 0, 1 1))'), _POINT_DIMENSION).wkt
+    'MULTIPOINT (-1 -1, 0 0, 1 1, 2 2, 3 3)'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (LINESTRING(-1 -1, 0 0), GEOMETRYCOLLECTION (LINESTRING(1 1, 2 2), GEOMETRYCOLLECTION (POINT(3 3))), LINESTRING(0 0, 1 1))'), _LINE_DIMENSION).wkt
+    'MULTILINESTRING ((-1 -1, 0 0), (1 1, 2 2), (0 0, 1 1))'
+    >>> _filter_geom_types(shapely.wkt.loads('GEOMETRYCOLLECTION (POLYGON((-2 -2, -2 2, 2 2, 2 -2, -2 -2)), GEOMETRYCOLLECTION (LINESTRING(1 1, 2 2), GEOMETRYCOLLECTION (POLYGON((3 3, 0 0, 1 0, 3 3)))), LINESTRING(0 0, 1 1))'), _POLYGON_DIMENSION).wkt
+    'MULTIPOLYGON (((-2 -2, -2 2, 2 2, 2 -2, -2 -2)), ((3 3, 0 0, 1 0, 3 3)))'
+    """
+
+    # flatten the geometries, and keep the parts with the
+    # dimension that we want. each item in the parts list
+    # should be a single (non-multi) geometry.
+    parts = []
+    for g in _flatten_geoms(shape):
+        if _geom_dimensions(g) == keep_dim:
+            parts.append(g)
+
+    if len(parts) == 0:
+        # the only way we can signal an empty geometry, as
+        # MultiPoint([]) throws an exception.
+        return GeometryCollection()
+
+    elif len(parts) == 1:
+        # return the singular geometry
+        return parts[0]
+
+    else:
+        # try to make a multi-geometry of the desirect type
+        if keep_dim == _POINT_DIMENSION:
+            # not sure why the MultiPoint constructor wants
+            # its coordinates differently from MultiPolygon
+            # and MultiLineString...
+            coords = []
+            for p in parts:
+                coords.extend(p.coords)
+            return MultiPoint(coords)
+
+        elif keep_dim == _LINE_DIMENSION:
+            return MultiLineString(parts)
+
+        elif keep_dim == _POLYGON_DIMENSION:
+            return MultiPolygon(parts)
+
+        else:
+            raise Exception("Unknown dimension %d in _filter_geom_types" % keep_dim)
```

Comment:
Not that it particularly matters here, but would a `ValueError` be more appropriate?